[sparse] add proper path for cloning sparse tensors

aartbik opened the pull request:
The code does the right thing (rather than crashing). This is a small step towards https://github.com/pytorch/pytorch/issues/117188

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

aartbik committed:
[sparse] add proper paths for non-strided tensors (mainly sparse)

The code does the right thing (rather than crashing).
This is a small step towards https://github.com/pytorch/pytorch/issues/117188

aartbik commented:
@pearu to review

pearu commented on torch/_dynamo/output_graph.py:
```diff
@@ -633,8 +633,9 @@ def bind_symint(s, prop):
                 bind_symint(s, TensorPropertySource(src, TensorProperty.SIZE, i))
-            for i, s in enumerate(t.stride()):
-                bind_symint(s, TensorPropertySource(src, TensorProperty.STRIDE, i))
+            if t.layout is torch.strided:
```
This if-block should capture `storage_offset` as well because `storage_offset` is stride layout specific property.

aartbik replied:
I will leave this fix than for a later PR when it is easier to test

pearu reviewed (commented):
Do we have a reproducer of the crash?

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
added test for clone path

aartbik commented:
> Do we have a reproducer of the crash?

I added a test for this (by calling clone directly).

Note that I added this in the test file that will eventually also contain the propagation tests (see other PRs)

aartbik committed:
no skip

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
addressed Pearu's feedback

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
formatting

aartbik commented:
@pearu is this (smaller) subset of the larger PR good to go?

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
break 80 col length

pearu commented on test/test_sparse_prop.py:
```diff
@@ -0,0 +1,66 @@
+    @parametrize("dtype", SPARSE_DTYPES)
+    @parametrize("itype", SPARSE_ITYPES)
+    @parametrize("layout", SPARSE_LAYOUTS)
+    def test_copy(self, dtype, itype, layout):
```
Could you move this test method, say, to test/dynamo/test_misc.py as "test_clone_input" that might be more appropriate place for testing dynamo specific utils.

aartbik replied:
sounds good

pearu commented on torch/_dynamo/utils.py:
```diff
@@ -774,6 +774,40 @@ def torch_clone(x):
+            return torch.sparse_coo_tensor(
+                torch_clone(x._indices()), torch_clone(x._values()), x.shape
+            )
+        elif x.layout is torch.sparse_csr:
+            return torch.sparse_csr_tensor(
+                torch_clone(x.crow_indices()),
+                torch_clone(x.col_indices()),
+                torch_clone(x.values()),
+                x.shape
+            )
```
Let's reduce LOC here, say, via
```suggestion
        elif is_sparse_compressed(x):
            if x.layout in {torch.sparse_csr, torch.sparse_bsr}:
                 compressed_indices = x.crow_indices()
                 plain_indices = x.col_indices()
            else:
                 compressed_indices = x.ccol_indices()
                 plain_indices = x.row_indices()
            return torch.sparse_compressed_tensor(
                torch_clone(compressed_indices),
                torch_clone(plain_indices),
                torch_clone(x.values()),
                x.shape,
                layout=x.layout
            )
```
where
```python
from torch._subclasses.meta_utils import is_sparse_compressed
```

aartbik replied:
oh, I love it! very good suggestion, thanks!

pearu commented on torch/_dynamo/utils.py:
```diff
@@ -774,6 +774,40 @@ def torch_clone(x):
+                torch_clone(x.values()),
+                x.shape
+            )
+        elif x.layout is torch.sparse_csc:
+            return torch.sparse_csc_tensor(
+                torch_clone(x.ccol_indices()),
+                torch_clone(x.row_indices()),
+                torch_clone(x.values()),
+                x.shape
+            )
+        elif x.layout is torch.sparse_bsr:
+            return torch.sparse_bsr_tensor(
+                torch_clone(x.crow_indices()),
+                torch_clone(x.col_indices()),
+                torch_clone(x.values()),
+                x.shape
+            )
+        elif x.layout is torch.sparse_bsc:
+            return torch.sparse_bsc_tensor(
+                torch_clone(x.ccol_indices()),
+                torch_clone(x.row_indices()),
+                torch_clone(x.values()),
+                x.shape
+            )
+
```
```suggestion
```

pearu reviewed (changes requested):
I have a couple of suggestions:
- move the dynamo specific test method to dynamo tests
- reduce LOC
- pass on the coalesce state to the cloned tensor

aartbik commented on torch/_dynamo/utils.py:
```diff
@@ -774,6 +774,40 @@ def torch_clone(x):
+        # Handle sparse storage (no stride).
+        if x.layout is torch.sparse_coo:
+            return torch.sparse_coo_tensor(
+                torch_clone(x._indices()), torch_clone(x._values()), x.shape
```
right you are! done

aartbik commented on test/test_sparse_prop.py:
```diff
@@ -0,0 +1,66 @@
+            # Invoke the dynamo clone input method directly.
+            sparse_copy = torch._dynamo.utils.clone_input(sparse_input)
+            # Make sure copy is successful.
+            self.assertEqual(sparse_input.layout, sparse_copy.layout)
+            self.assertEqual(sparse_input.shape, sparse_copy.shape)
```
agreed, a lot simpler ;-)

aartbik committed:
reviewer feedback

aartbik committed:
Merge branch 'pytorch:main' into bok

pearu reviewed (approved):
LGTM! Thanks, @aartbik!

aartbik committed:
fix lint

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
edit

cpuhrsch commented:
@pytorchbot merge

cpuhrsch commented:
@pytorchbot merge

cpuhrsch commented:
@pytorchbot merge

cpuhrsch commented:
Looks like there's a lint issue. You can install the linter via https://github.com/pytorch/pytorch/wiki/lintrunner

aartbik committed:
Update test_misc.py
